runtime-light/stdlib/array/array-functions.h

```diff
@@ -60,9 +65,17 @@ array<T> f$array_filter_by_key(const array<T> &a, const T1 &callback) noexcept {
   php_critical_error("call to unsupported function");
 }
 
-template<class T, class CallbackT, class R = typename std::invoke_result_t<std::decay_t<CallbackT>, T>>
-array<R> f$array_map(const CallbackT &callback, const array<T> &a) {
-  php_critical_error("call to unsupported function");
+template<class A, std::invocable<A> F, class R = async_function_unwrapped_return_type_t<F, A>>
+task_t<array<R>> f$array_map(F f, array<A> arr) {
+  array<R> result{arr.size()};
+  for (const auto &it : arr) {
+    if constexpr (is_async_function_v<F, A>) {
+      result.set_value(it.get_key(), co_await std::invoke(f, it.get_value()));
+    } else {
+      result.set_value(it.get_key(), std::invoke(f, it.get_value()));
+    }
+  }
+  co_return result;
```

Comment:
But it forces `array_map` to be always async, isn't it?
Do we really want it to be async?

Reply:
Yes, `array_map` is async now according to its signature. So, in case of async callback it may stop its execution. And in case of a sync callback it will execute synchronously.

Reply:
But in any case it would be coroutine. Is it intended?

Reply:
Yep, it is

Reply:
But it means we need extra allocation for coroutine frame each time, isn't it?
It's OK for `array_map` as it allocates result inside, but let's see at `array_reduce`.
It doesn't allocate extra memory, but if we gonna implement it like that, than we would have extra useless allocations. Is it possible to have a template overload such that returns `T` instead of `task_t<T>`?
If yes, how hard to implement this?

If it's hard, we should leave a comment about it at least

Reply:
In general, we can't overload a function based on its return type. But it can be done with some hacks. I think it's better to have a clean implementation and apply those hacks only if we see that we have a performance gap

Reply:
After discussion we've decided to leave a comment